src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs

```diff
@@ -0,0 +1,243 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
+{
+    /// <summary>
+    /// provides message ordering and deduplication of request messages (operations or lock requests)
+    /// that are sent to entities, from other entities, or from orchestrations.
+    /// </summary>
+    [JsonObject(MemberSerialization.OptIn)]
+    internal class MessageSorter
+    {
+        // don't update the reorder window too often since the garbage collection incurs some overhead.
+        private static readonly TimeSpan MinIntervalBetweenCollections = TimeSpan.FromSeconds(10);
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public Dictionary<string, DateTime> Sent { get; set; }
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public Dictionary<string, ReceiveBuffer> Received { get; set; }
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public DateTime ReceiveHorizon { get; set; }
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public DateTime SendHorizon { get; set; }
+
+        public int NumberBufferedRequests =>
+            this.Received?.Select(kvp => kvp.Value.Buffered?.Count ?? 0).Sum() ?? 0;
+
+        /// <summary>
+        /// Called on the sending side, to fill in timestamp and predecessor fields.
+        /// </summary>
+        public void LabelOutgoingMessage(RequestMessage message, string destination, DateTime now, TimeSpan reorderWindow)
+        {
+            DateTime timestamp = now;
+
+            if (this.SendHorizon + reorderWindow + MinIntervalBetweenCollections < now)
```

Comment:
It can make a big difference for high-throughput situations. 

For example, consider that an entity sends 1000 messages in a batch. Without `MinIntervalBetweenCollection` we would then execute the collection code 1000 times. Since the collection code has an internal loop (`foreach (var kvp in this.Sent)` this can get expensive very quick. The `MinIntervalBetweenCollection` ensures we are not executing that code more than once every 10 seconds.

Reply:
Makes sense. Thanks for explaining!